gossip/src/crds_data.rs

```diff
@@ -0,0 +1,718 @@
+        // the mainnet crds table.
+        match kind {
+            0 => CrdsData::ContactInfo(ContactInfo::new_rand(rng, pubkey)),
+            // Index for LowestSlot is deprecated and should be zero.
+            1 => CrdsData::LowestSlot(0, LowestSlot::new_rand(rng, pubkey)),
+            2 => CrdsData::LegacySnapshotHashes(LegacySnapshotHashes::new_rand(rng, pubkey)),
+            3 => CrdsData::AccountsHashes(AccountsHashes::new_rand(rng, pubkey)),
+            4 => CrdsData::Version(Version::new_rand(rng, pubkey)),
+            5 => CrdsData::Vote(rng.gen_range(0..MAX_VOTES), Vote::new_rand(rng, pubkey)),
+            6 => CrdsData::RestartLastVotedForkSlots(RestartLastVotedForkSlots::new_rand(
+                rng, pubkey,
+            )),
+            7 => CrdsData::RestartHeaviestFork(RestartHeaviestFork::new_rand(rng, pubkey)),
+            _ => CrdsData::EpochSlots(
+                rng.gen_range(0..MAX_EPOCH_SLOTS),
+                EpochSlots::new_rand(rng, pubkey),
+            ),
+        }
+    }
+
+    pub(crate) fn wallclock(&self) -> u64 {
+        match self {
+            CrdsData::LegacyContactInfo(contact_info) => contact_info.wallclock(),
+            CrdsData::Vote(_, vote) => vote.wallclock,
+            CrdsData::LowestSlot(_, obj) => obj.wallclock,
+            CrdsData::LegacySnapshotHashes(hash) => hash.wallclock,
+            CrdsData::AccountsHashes(hash) => hash.wallclock,
+            CrdsData::EpochSlots(_, p) => p.wallclock,
+            CrdsData::LegacyVersion(version) => version.wallclock,
+            CrdsData::Version(version) => version.wallclock,
+            CrdsData::NodeInstance(node) => node.wallclock,
+            CrdsData::DuplicateShred(_, shred) => shred.wallclock,
+            CrdsData::SnapshotHashes(hash) => hash.wallclock,
+            CrdsData::ContactInfo(node) => node.wallclock(),
+            CrdsData::RestartLastVotedForkSlots(slots) => slots.wallclock,
+            CrdsData::RestartHeaviestFork(fork) => fork.wallclock,
+        }
+    }
+
+    pub(crate) fn pubkey(&self) -> Pubkey {
+        match &self {
+            CrdsData::LegacyContactInfo(contact_info) => *contact_info.pubkey(),
+            CrdsData::Vote(_, vote) => vote.from,
+            CrdsData::LowestSlot(_, slots) => slots.from,
+            CrdsData::LegacySnapshotHashes(hash) => hash.from,
+            CrdsData::AccountsHashes(hash) => hash.from,
+            CrdsData::EpochSlots(_, p) => p.from,
+            CrdsData::LegacyVersion(version) => version.from,
+            CrdsData::Version(version) => version.from,
+            CrdsData::NodeInstance(node) => node.from,
+            CrdsData::DuplicateShred(_, shred) => shred.from,
+            CrdsData::SnapshotHashes(hash) => hash.from,
+            CrdsData::ContactInfo(node) => *node.pubkey(),
+            CrdsData::RestartLastVotedForkSlots(slots) => slots.from,
+            CrdsData::RestartHeaviestFork(fork) => fork.from,
+        }
+    }
+}
```

Comment:
unrelated to this exact PR, but why are `pubkey` and `wallclock` private in `ContactInfo` and `LegacyContactInfo` but public in the other `CrdsData` types

Reply:
A lot of these variants are defined in this same module.
So even if they were private, this code could still access them directly.